File: flink-sql/src/main/java/org/apache/flink/table/plan/rules/DynamicDataStreamCalcRule.java
package org.apache.flink.table.plan.rules;

import static org.apache.flink.table.plan.nodes.DynamicFlinkConventions.DYNAMIC_DATA_STREAM;

import org.apache.calcite.plan.RelOptRule;
import org.apache.calcite.plan.RelTraitSet;
import org.apache.calcite.rel.RelNode;
import org.apache.calcite.rel.convert.ConverterRule;
import org.apache.flink.table.plan.nodes.FlinkConventions;
import org.apache.flink.table.plan.nodes.datastream.DynamicDataStreamCalc;
import org.apache.flink.table.plan.nodes.logical.FlinkLogicalCalc;

public class DynamicDataStreamCalcRule extends ConverterRule {

  public static DynamicDataStreamCalcRule INSTANCE = new DynamicDataStreamCalcRule();

  private DynamicDataStreamCalcRule() {
    super(FlinkLogicalCalc.class, FlinkConventions.LOGICAL(), DYNAMIC_DATA_STREAM, "DynamicDataStreamCalcRule");
  }

  @Override
  public RelNode convert(RelNode relNode) {
    FlinkLogicalCalc calc = (FlinkLogicalCalc) relNode;
    RelTraitSet traitSet = relNode.getTraitSet().replace(DYNAMIC_DATA_STREAM);
    RelNode convertInput = RelOptRule.convert(calc.getInput(), DYNAMIC_DATA_STREAM);

    return new DynamicDataStreamCalc(relNode.getCluster(),
        traitSet,
        convertInput,
        calc.getProgram());
  }
}

File: flink-extended-sql/src/main/scala/org/apache/flink/table/plan/rules/datastream/DynamicDataStreamCalcRule.scala
package org.apache.flink.table.plan.rules.datastream

import org.apache.calcite.plan.{RelOptRule, RelTraitSet}
import org.apache.calcite.rel.RelNode
import org.apache.calcite.rel.convert.ConverterRule
import org.apache.flink.table.plan.nodes.FlinkConventions
import org.apache.flink.table.plan.nodes.datastream.DynamicDataStreamCalc
import org.apache.flink.table.plan.nodes.logical.FlinkLogicalCalc
import org.apache.flink.table.plan.schema.RowSchema

class DynamicDataStreamCalcRule extends ConverterRule (
  classOf[FlinkLogicalCalc],
  FlinkConventions.LOGICAL,
  FlinkConventions.DATASTREAM,
  "DynamicDataStreamCalcRule") {

  override def convert(rel: RelNode): RelNode = {
    val calc: FlinkLogicalCalc = rel.asInstanceOf[FlinkLogicalCalc]
    val traitSet: RelTraitSet = rel.getTraitSet.replace(FlinkConventions.DATASTREAM)
    val convInput: RelNode = RelOptRule.convert(calc.getInput, FlinkConventions.DATASTREAM)

    new DynamicDataStreamCalc(
      rel.getCluster,
      traitSet,
      convInput,
      new RowSchema(convInput.getRowType),
      new RowSchema(rel.getRowType),
      calc.getProgram,
      "DynamicDataStreamCalcRule")
  }

}


object DynamicDataStreamCalcRule {

  val INSTANCE = new DynamicDataStreamCalcRule

}